docs/en/feature/multi-instance-feature.md

```diff
@@ -0,0 +1,113 @@
+
+A multi-instance feature is a feature which have multiple instances in the redux state. A state for a multi-feature instance, instead of being stored directly in the app state by the feature name key, like this:
+
+```
+{
+  ...app state,
+  [featureName]: featureState,
+}
+```
+is now stored using an assigned to the feature instance unique ID (also called `instanceKey`):
+```
+{
+  ...app state,
+  [featureName]: {
+    [featureInstanceID]: featureInstanceState,
+    [anotherFeatureInstanceID]: anotherFeatureInstanceState,
+    ...and so on
+  }
+}
+```
+## How to use a multi-instance feature:
+Let's say we have a widget in our app which is represented by a `widget` feature. A user can add as many instances (the feature containers) of that widget on the page as he wants. The widget displays some data fetched from a server. The user can also configure the widget to change the data that gets fetched and displayed.
+
+So in this case every container with the widget would need to be able to use its own redux state, which can be achieved by making the feature multi-instance.
+
+To make a feature multi-instance you should:
+
+1. Wrap a feature's state interface (written as usual) with the IMultiInstanceState inside the IAppReduxState interface.
+   ```
+   export interface IAppReduxState {
+    ...
+    widget: IMultiInstanceState<features.widget.namespace.IReduxState>;
+   }
+   ```
+
+2. Write selectors for a multi-instance feature with the feature state as its argument instead of the app state.
+   ```
+   import * as NS from '../namespace';
+
+   export function selectWidgetData(state: NS.IReduxState): IWidgetData {
+     return state.data.widgetData;
+   }
+   ```
+   Action creators are written as usual.
+
+3. Use the `multiConnect` HOC instead of the regular redux `connect` in the feature's containers. `multiConnect` takes the following arguments:
+   - path to the feature state (array of strings);
+   - initial feature state;
+   - mapStateToProps function (the first argument is the feature state, as the multi-instance feature selectors are taking the feature state as its argument instead of the app state; the other arguments are the same as in the regular `connect`, starting from appState);
+      ```
+      function mapStateToProps(state: IReduxState, appState: IAppReduxState): IStateProps {
+        return {
+          widgetData: selectors.selectWidgetData(state),
+          UITheme: configServiceSelectors.selectUITheme(appState),
+        };
+      }
+      ```
+   - mapDispatchToProps function (written just like one for the regular `connect`);
+
+   A component wrapped with `multiConnect` also have the `instanceKey` prop. If you want to make use of it you should use the `IMultiConnectProps` in props type. `instanceKey` is either generated automatically by the `multiConnect` or passed explicitly to the component in case you need multiple containers to use the same redux state branch.
```

Comment:
to the component in case you need multiple containers to use the same redux state branch. ->
to the component in the case when you need multiple containers to use the same redux state branch.

И вообще на костыль похоже

Reply:
Про второе [ты не прав](https://dictionary.cambridge.org/grammar/british-grammar/conditionals-and-wishes/in-case-of).

Reply:
В первом случае скорее запятая.
Про второе - я бы там вообще заменил на простой if, там in case как-то не очень в тему на самом деле.
Как думаешь?

Reply:
> В первом случае скорее запятая.

ок
> Про второе - я бы там вообще заменил на простой if, там in case как-то не очень в тему на самом деле.
Как думаешь?

Норм
> Про костыль не понял, о чем речь.

Да я снова про АПИ мультиконнекта. Забей)